### DISCUSSION TITLE
Filer gRPC - download file

### DISCUSSION THREAD

**Discussion by mbuczko** [seaweedfs/seaweedfs#2403] [General]
Filer proto exposes `DownloadToLocal` rpc - I thought I could use it to download given file locally but I'm getting back _no entry is found in filer store_ instead. However, the same file is listed when using `ListEntries` in requested directory, so I believe I provide the name and directory correctly. Any hint what I missed here? Thanks!

**Comment by chrislusf** [comment]
the API `DownloadToLocal` is used to download from remote object store to local cluster. I will rename this API to avoid future confusion.

for file downloading, use HTTP API.

**Comment by mbuczko** [comment]
Ah, I've got a bit confused by the naming :)

Thanks for explanation.

**Reply by mbuczko** [reply]
hm.. I need a _file_id_ to fetch file using HTTP API, right? if so, is there any reason why `SubscribeMetadata` event returns entry without `file_id`:
```json
{
  "directory": "/pic",
  "event_notification": {
    "signatures": [
      -971019148
    ],
    "old_entry": null,
    "new_entry": {
      "chunks": [
        {
          "file_id": "",
          "offset": "0",
          "size": "1750417",
          "mtime": "1635544790668702085",
          "e_tag": "g8XJ7uYvRIPVh35T+BfiWQ==",
          "source_file_id": "",
          "fid": {
            "volume_id": 13,
            "file_key": "265",
            "cookie": 4180831274
          }, 
          ...
```
I guess I can assemble file_id directly from _fid_? 

By the way, is there any case where this event would contain more than 1 chunk for file being uploaded?

Edit: a bit of context. I'm trying to listen for newly uploaded files to extract their meta data. To do this, I subscribe to `SubscribeMetadata` and fetching file locally based on incoming events.

**Reply by chrislusf** [reply]
"I guess I can assemble file_id directly from fid?" Yes.
"is there any case where this event would contain more than 1 chunk for file being uploaded?". Yes, if the file is larger than default chunk size limit.

You can read the file content via HTTP API.

**Reply by mbuczko** [reply]
> You can read the file content via HTTP API.

right, I didn't notice it simplifies all these things. thanks once again, that answers all my questions :)
